shell/platform/windows/flutter_window.cc

```diff
@@ -0,0 +1,286 @@
+  // so
+  // that this isn't a tangle of references.
+  auto messenger = std::make_unique<FlutterDesktopMessenger>();
+  message_dispatcher_ =
+      std::make_unique<flutter::IncomingMessageDispatcher>(messenger.get());
+  messenger->engine = engine_;
+  messenger->dispatcher = message_dispatcher_.get();
+
+  window_wrapper_ = std::make_unique<FlutterDesktopWindow>();
+  window_wrapper_->window = this;
+
+  plugin_registrar_ = std::make_unique<FlutterDesktopPluginRegistrar>();
+  plugin_registrar_->messenger = std::move(messenger);
+  plugin_registrar_->window =
+      window_wrapper_.get();  // state->window_wrapper.get();
+
+  internal_plugin_registrar_ =
+      std::make_unique<flutter::PluginRegistrar>(plugin_registrar_.get());
+
+  // Set up the keyboard handlers.
+  auto internal_plugin_messenger = internal_plugin_registrar_->messenger();
+  keyboard_hook_handlers_.push_back(
+      std::make_unique<flutter::KeyEventHandler>(internal_plugin_messenger));
+  keyboard_hook_handlers_.push_back(
+      std::make_unique<flutter::TextInputPlugin>(internal_plugin_messenger));
+  platform_handler_ = std::make_unique<flutter::PlatformHandler>(
+      internal_plugin_messenger, this);
+
+  auto state = std::make_unique<FlutterDesktopWindowControllerState>();
+  state->engine = engine_;
+  state->window = nullptr;  // null here as the caller owns the uniqueptr we'll
+                            // ultimately want to return
+  state->window_wrapper =
+      std::make_unique<FlutterDesktopWindow>();  // a different window wrapper
+                                                 // for same window pointer
+                                                 // hence make_unique again
+  state->window_wrapper->window = this;
+  return state.release();
+}
+
+FlutterDesktopPluginRegistrarRef FlutterWindow::GetRegistrar() {
+  return plugin_registrar_.get();
+}
+
+void FlutterWindow::HandleMessage(const FlutterDesktopMessage& message,
+                                  std::function<void(void)> input_block_cb,
+                                  std::function<void(void)> input_unblock_cb) {
+  message_dispatcher_->HandleMessage(message, input_block_cb, input_unblock_cb);
+}
+
+// When DesktopWindow notifies that a WM_Size message has come in
+// let FlutterEngine know about the new size.
+void FlutterWindow::OnResize(unsigned int width, unsigned int height) {
+  current_width_ = width;
+  current_height_ = height;
+  SendWindowMetrics();
+}
+// When DesktopWindow notifies that a WM_Size message has come in
+// let FlutterEngine know about the new size.
+void FlutterWindow::OnResize() {
+  SendWindowMetrics();
+}
+
+// When DesktopWindow notifies that the DPI has changed
+// handle the change.
+void FlutterWindow::OnDpiScale(unsigned int dpi) {
+  current_dpi_ = dpi;
+
+  // No need to SendWindowMetrics here as OnResize will be called right after
+  // DPI change
+}
+
+void FlutterWindow::OnPointerMove(double x, double y) {
+  SendPointerMove(x, y);
```

Comment:
What's the reason for the extra level of indirection in all of these methods? Each one seems to be calling a private method that's only ever used once, and has the same basic meaning/semantics as the public method.

Reply:
Take a look at the latest version, now checking if event sending is enabled.